cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,230 @@
+
+For example, our ToDo application could be updated to have the following "name" parameter:
+
+```bicep
+@minLength(1)
+@maxLength(64)
+@description('The name used as a prefix for the resource group which holds all the resources for the ToDo application. The suffix `-rg` will automatically be added.')
+param name string
+```
+
+Which provides template specific information on how the parameter is used. It also ensures that these templates behave nicely with other deployment tools (like creating a deployment from the portal or using `az` directly).
+
+We also allow some `azd` specific metadata to be applied to properties:
+
+```bicep
+@description('The Azure location to use for all resources.')
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+The `type` property of the `azd` metadata object denotes that this value should be set to an Azure location name (e.g. "westus2") and is used by `azd` to offer a "pick a location from this list of all Azure locations). This composes with other validation decorators that you can apply, for example `@allowed()` which can be used to restrict the set of locations offered to the user to pick from (for cases where a template uses services that are not deployed to all regions):
+
+```bicep
+@description('The Azure location to use for all resources.')
+@allowed([
+    'westus2'
+    'centralus'
+])
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+To start, we support the following types:
+
+`location`: The value should be the short name of an Azure location. All locations are supported by default, but the exiting `@allowed` may restrict this set.
+`current-principal-id`: The value should be the id of the authenticated principal.
+
+Configuration values (including the environment name and id) are **not** stored in the `.env` file.
+
+Deployment outputs are also stored in the `config.json` file under the "outputs" key (a sibling of "properties") and updated on each deployment. They are written to the `.env` file as before, as well.
+
+The intermediate `.parameters.json` files are no longer written in the `.azure\<environment-name>` directory instead they are written in a temporary directory (and when we stop depending on `az` for deployments will not be persisted to disk)
+
+When creating names for deployments, we name them using the combination of the environment name and ID (e.g. `default-1539e0b8-ac32-440c-ae1f-d131512563dd`), if this would generate a deployment name which is too long, we truncate the environment name but keep the full id). This ensures the deployment names don't clash between two unrelated projects even if they have the same environment name (which will now be common, since we have a default value and no longer force the user into picking a new environment).
```

Comment:
> I'm thinking about what the resource group/resources will look like when they go to the portal or look at things in the VS Code siderail. I'm imagining that if they have a couple projects all starting with "default-..." then the names are pretty hard to parse (because they'd be relying on remembering/knowing the random ID to figure out which resource group belongs to which project etc.)

This is subtle, but the name of the deployment has no impact on the resources provisioned by that deployment.  It's simply a named object that is used to hold information about the deployment that is progressing.

The name ends up being visible to users if they look at the "deployments" tab of their subscription (or when we target resource groups, looking at the "deployments" tab of the resource group) and in the URL we print during a `provision` to view progress on the Azure portal, but otherwise is not something developers would think about day to day.

If we adopt this model, our templates will still take a `name` parameter, and that name will be used to create the resource group and the user has full control over this name (just as they have full control over any other parameter of their deployment)

Having said that, I can see the value of including the project name in this string - If we adopt `<project-name>-<env-name>-<env-id>` we'll end up with names that may be too longer for a deployment (the maximum length is 64 characters, the UUID takes 36 characters, "default" is 7 characters and the two hypens between these tokens and project names are usually longer strings right now like `todo-nodejs-mongo-aca` so we'd have to do some level of shortening.  `<project-name>-<environment-id>` is nice because it is shorter and less likely to have to be truncated but it does make the names a bit harder to interpret.

> We'd have to prompt for the project name or just infer it from the template/directory name.

There is actually a `name` property in azure.yaml today which we can use as the project name (and if you run `azd init` without a template the name defaults to the name of the directory you are initializing into)

Reply:
(more of an observation to myself) We may not need to include the entire id in the deployment name, but we should make it possible to collect all deployments for an environment by its ID somehow (I believe you can associate a tag with the deployment object, but this may not be possible with the CLI)

I suspect we will also want to consider adding a timestamp to the name or some other value that changes per deployment - not having a separate deployment object per deployment seems like it will be frustrating long term (we already see this a bit with some of the "get deployments operations to show progress" stuff we do since we can end up seeing the previous deployment operations deployment since it's URI and the new deployment URI end up being the same.  If we had separate names this would no longer be an issue.

Reply:
Adjusted based on the discussion here.